### PR TITLE
Fix DispatchMesh with groupshared payload.

### PR DESCRIPTION
**Pull request by tex3d**

### CODE REVIEW

File: lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp [hunk 1]
```diff
@@ -777,6 +778,200 @@ static unsigned getNestedLevelInStruct(const Type *ty) {
...
+/// Returns first GEP index that indexes a struct member, or 0 otherwise.
+/// Ignores initial ptr index.
+static unsigned FindFirstStructMemberIdxInGEP(GEPOperator *GEP) {
+  StructType *ST = dyn_cast<StructType>(
+    GEP->getPointerOperandType()->getPointerElementType());
+  int index = 1;
+  for (auto it = gep_type_begin(GEP), E = gep_type_end(GEP); it != E;
+       ++it, ++index) {
+    if (ST) {
+      DXASSERT(!HLMatrixType::isa(ST) && !dxilutil::IsHLSLObjectType(ST),
```
Comment (python3kgae) [inline lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp]: Why this will not happen?

File: lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp [hunk 2]
```diff
@@ -777,6 +778,200 @@ static unsigned getNestedLevelInStruct(const Type *ty) {
...
+    return Ptr;
+
+  // If type changed, replace value, otherwise casting may still
+  // require a rewrite of the calls.
+  Value *NewPtr = Ptr;
+  if (Ty != NewTy) {
+    // TODO: Transfer type annotation
+    DxilStructAnnotation *pOldAnnotation = TypeSys.GetStructAnnotation(cast<StructType>(Ty));
+    if (pOldAnnotation) {
+
```
Comment (python3kgae) [inline lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp]: Something lost here?

File: lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp (hunk 1 above)
Comment (tex3d) [reply lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp]: Should not have GEP indexing into matrix or object struct unless there's an internal bug.

File: lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp (hunk 2 above)
Comment (tex3d) [reply lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp]: Shoot, forgot to clean this up and provide typesys to GetLoweredUDT, where type annotation is copied.  I'll fix that.

**Comment by tex3d** [comment]
> Could we have a multi-dim array test case?

Actually there is.  The as-groupshared-payload-matrix.hlsl test has array of GSStruct, which has array of MeshPayload struct, and since SROA succeeds on GSStruct, it turns the MeshPayload member into a 2-dim array, requiring the MultiDimToSingleDim change to succeed, since it's now translating a multi-dim array of struct to a single-dim array of struct.

**Pull request merged**